File: 121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  int maxProfit(vector<int>& prices) {
    int max_profit = 0;
    int min_price = INT_MAX;
    for (int price : prices) {
      min_price = min(min_price, price);
      max_profit = max(max_profit, (price - min_price));
    }

    return max_profit;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> vec;
  vec = {};
  assert(s.maxProfit(vec) == 0);
  vec = {1};
  assert(s.maxProfit(vec) == 0);
  vec = {7, 1, 5, 3, 6, 4};
  assert(s.maxProfit(vec) == 5);
  vec = {7, 6, 4, 3, 1};
  assert(s.maxProfit(vec) == 0);
  return 0;
}

File: 121.best-time-to-buy-and-sell-stock.rs
/*
 * @lc app=leetcode id=121 lang=rust
 *
 * [121] Best Time to Buy and Sell Stock
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[macro_export]
macro_rules! list_node {
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let mut node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min_price = i32::MAX;
        let mut max_profit = i32::MIN;
        for price in prices {
            min_price = std::cmp::min(min_price, price);
            max_profit = std::cmp::max(max_profit, price - min_price);
        }
        max_profit
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::max_profit(vec![7, 1, 5, 3, 6, 4]), 5);
}

#[test]
fn test2() {
    assert_eq!(Solution::max_profit(vec![7, 6, 4, 3, 1]), 0);
}
